File: backend/src/main/java/com/bhojpurilang/interpreter/Interpreter.java
package com.bhojpurilang.interpreter;

import com.bhojpurilang.ast.*;

import java.util.List;

public class Interpreter {
    private final Environment environment = new Environment();
    private final StringBuilder output = new StringBuilder();

    public String execute(Program program) {
        for (Statement stmt : program.getStatements()) {
            executeStatement(stmt);
        }
        return output.toString();
    }

    private void executeStatement(Statement stmt) {
        if (stmt instanceof VariableDeclaration) {
            VariableDeclaration varDecl = (VariableDeclaration) stmt;
            Object val = evaluate(varDecl.getInitializer());
            environment.define(varDecl.getIdentifier(), val);
        } else if (stmt instanceof Assignment) {
            Assignment assign = (Assignment) stmt;
            Object val = evaluate(assign.getValue());
            environment.assign(assign.getIdentifier(), val, assign.getLine());
        } else if (stmt instanceof PrintStatement) {
            PrintStatement printStmt = (PrintStatement) stmt;
            Object val = evaluate(printStmt.getExpression());
            if (output.length() > 0) {
                output.append("\n");
            }
            output.append(stringify(val));
        } else if (stmt instanceof IfStatement) {
            IfStatement ifStmt = (IfStatement) stmt;
            Object condVal = evaluate(ifStmt.getCondition());
            if (isTruthy(condVal)) {
                for (Statement s : ifStmt.getThenBranch()) {
                    executeStatement(s);
                }
            } else {
                for (Statement s : ifStmt.getElseBranch()) {
                    executeStatement(s);
                }
            }
        } else if (stmt instanceof WhileStatement) {
            WhileStatement whileStmt = (WhileStatement) stmt;
            int iterations = 0;
            int MAX_ITERATIONS = 100000; // Protection against infinite loops

            while (isTruthy(evaluate(whileStmt.getCondition()))) {
                iterations++;
                if (iterations > MAX_ITERATIONS) {
                    throw new RuntimeException("Runtime Error at line " + whileStmt.getLine() + ": Infinite loop detected (exceeded 100,000 iterations)");
                }
                for (Statement s : whileStmt.getBody()) {
                    executeStatement(s);
                }
            }
        }
    }

    private Object evaluate(Expression expr) {
        if (expr instanceof LiteralExpression) {
            return ((LiteralExpression) expr).getValue();
        } else if (expr instanceof VariableExpression) {
            VariableExpression varExpr = (VariableExpression) expr;
            return environment.get(varExpr.getName(), varExpr.getLine());
        } else if (expr instanceof BinaryExpression) {
            BinaryExpression bin = (BinaryExpression) expr;
            Object left = evaluate(bin.getLeft());
            Object right = evaluate(bin.getRight());
            String op = bin.getOperator();
            int line = bin.getLine();

            switch (op) {
                case "+":
                    if (left instanceof String || right instanceof String) {
                        return stringify(left) + stringify(right);
                    }
                    if (left instanceof Number && right instanceof Number) {
                        return ((Number) left).longValue() + ((Number) right).longValue();
                    }
                    throw new RuntimeException("Runtime Error at line " + line + ": Operands for '+' must be numbers or strings");

                case "-":
                    checkNumberOperands(op, left, right, line);
                    return ((Number) left).longValue() - ((Number) right).longValue();

                case "*":
                    checkNumberOperands(op, left, right, line);
                    return ((Number) left).longValue() * ((Number) right).longValue();

                case "/":
                    checkNumberOperands(op, left, right, line);
                    long denominator = ((Number) right).longValue();
                    if (denominator == 0) {
                        throw new RuntimeException("Runtime Error at line " + line + ": Division by zero");
                    }
                    return ((Number) left).longValue() / denominator;

                case "%":
                    checkNumberOperands(op, left, right, line);
                    long mod = ((Number) right).longValue();
                    if (mod == 0) {
                        throw new RuntimeException("Runtime Error at line " + line + ": Modulo by zero");
                    }
                    return ((Number) left).longValue() % mod;

                case "<":
                    checkNumberOperands(op, left, right, line);
                    return ((Number) left).longValue() < ((Number) right).longValue();

                case "<=":
                    checkNumberOperands(op, left, right, line);
                    return ((Number) left).longValue() <= ((Number) right).longValue();

                case ">":
                    checkNumberOperands(op, left, right, line);
                    return ((Number) left).longValue() > ((Number) right).longValue();

                case ">=":
                    checkNumberOperands(op, left, right, line);
                    return ((Number) left).longValue() >= ((Number) right).longValue();

                case "==":
                    return isEqual(left, right);

                case "!=":
                    return !isEqual(left, right);
            }
        }

        throw new RuntimeException("Runtime Error at line " + expr.getLine() + ": Cannot evaluate expression");
    }

    private void checkNumberOperands(String op, Object left, Object right, int line) {
        if (left instanceof Number && right instanceof Number) return;
        throw new RuntimeException("Runtime Error at line " + line + ": Operands for '" + op + "' must be integers");
    }

    private boolean isTruthy(Object object) {
        if (object == null) return false;
        if (object instanceof Boolean) return (Boolean) object;
        if (object instanceof Number) return ((Number) object).longValue() != 0;
        return true;
    }

    private boolean isEqual(Object a, Object b) {
        if (a == null && b == null) return true;
        if (a == null) return false;
        return a.equals(b);
    }

    private String stringify(Object object) {
        if (object == null) return "null";
        if (object instanceof Boolean) {
            return ((Boolean) object) ? "sahi" : "galat";
        }
        return object.toString();
    }
}

File: src/engine/interpreter.ts
import {
  ProgramNode,
  StatementNode,
  ExpressionNode,
  VariableDeclarationNode,
  AssignmentNode,
  PrintStatementNode,
  IfStatementNode,
  WhileStatementNode,
  BinaryExpressionNode,
  LiteralExpressionNode,
  VariableExpressionNode,
} from "./ast";

export class Environment {
  private values: Map<string, any> = new Map();

  public define(name: string, value: any): void {
    this.values.set(name, value);
  }

  public assign(name: string, value: any, line: number): void {
    if (this.values.has(name)) {
      this.values.set(name, value);
      return;
    }
    throw new Error(`Runtime Error at line ${line}: Unknown variable '${name}'`);
  }

  public get(name: string, line: number): any {
    if (this.values.has(name)) {
      return this.values.get(name);
    }
    throw new Error(`Runtime Error at line ${line}: Unknown variable '${name}'`);
  }

  public getAll(): Record<string, any> {
    const obj: Record<string, any> = {};
    this.values.forEach((val, key) => {
      obj[key] = val;
    });
    return obj;
  }
}

export class BhojpuriInterpreter {
  private environment: Environment = new Environment();
  private outputLogs: string[] = [];

  public execute(program: ProgramNode): { output: string; logs: string[]; environment: Record<string, any> } {
    this.outputLogs = [];
    this.environment = new Environment();

    for (const stmt of program.statements) {
      this.executeStatement(stmt);
    }

    return {
      output: this.outputLogs.join("\n"),
      logs: this.outputLogs,
      environment: this.environment.getAll(),
    };
  }

  private executeStatement(stmt: StatementNode): void {
    switch (stmt.kind) {
      case "VariableDeclaration": {
        const decl = stmt as VariableDeclarationNode;
        const val = this.evaluate(decl.initializer);
        this.environment.define(decl.identifier, val);
        break;
      }

      case "Assignment": {
        const assign = stmt as AssignmentNode;
        const val = this.evaluate(assign.value);
        this.environment.assign(assign.identifier, val, assign.line);
        break;
      }

      case "PrintStatement": {
        const printStmt = stmt as PrintStatementNode;
        const val = this.evaluate(printStmt.expression);
        this.outputLogs.push(this.stringify(val));
        break;
      }

      case "IfStatement": {
        const ifStmt = stmt as IfStatementNode;
        const cond = this.evaluate(ifStmt.condition);
        if (this.isTruthy(cond)) {
          for (const s of ifStmt.thenBranch) {
            this.executeStatement(s);
          }
        } else {
          for (const s of ifStmt.elseBranch) {
            this.executeStatement(s);
          }
        }
        break;
      }

      case "WhileStatement": {
        const whileStmt = stmt as WhileStatementNode;
        let iterations = 0;
        const MAX_ITERATIONS = 100000;

        while (this.isTruthy(this.evaluate(whileStmt.condition))) {
          iterations++;
          if (iterations > MAX_ITERATIONS) {
            throw new Error(`Runtime Error at line ${whileStmt.line}: Infinite loop detected (exceeded 100,000 iterations)`);
          }
          for (const s of whileStmt.body) {
            this.executeStatement(s);
          }
        }
        break;
      }

      default:
        throw new Error(`Runtime Error at line ${stmt.line}: Unknown statement type '${(stmt as any).kind}'`);
    }
  }

  private evaluate(expr: ExpressionNode): any {
    switch (expr.kind) {
      case "LiteralExpression":
        return (expr as LiteralExpressionNode).value;

      case "VariableExpression": {
        const varExpr = expr as VariableExpressionNode;
        return this.environment.get(varExpr.name, varExpr.line);
      }

      case "BinaryExpression": {
        const bin = expr as BinaryExpressionNode;
        const left = this.evaluate(bin.left);
        const right = this.evaluate(bin.right);
        const op = bin.operator;
        const line = bin.line;

        switch (op) {
          case "+":
            if (typeof left === "string" || typeof right === "string") {
              return this.stringify(left) + this.stringify(right);
            }
            if (typeof left === "number" && typeof right === "number") {
              return left + right;
            }
            throw new Error(`Runtime Error at line ${line}: Operands for '+' must be numbers or strings`);

          case "-":
            this.checkNumberOperands(op, left, right, line);
            return left - right;

          case "*":
            this.checkNumberOperands(op, left, right, line);
            return left * right;

          case "/":
            this.checkNumberOperands(op, left, right, line);
            if (right === 0) {
              throw new Error(`Runtime Error at line ${line}: Division by zero`);
            }
            return Math.floor(left / right);

          case "%":
            this.checkNumberOperands(op, left, right, line);
            if (right === 0) {
              throw new Error(`Runtime Error at line ${line}: Modulo by zero`);
            }
            return left % right;

          case "<":
            this.checkNumberOperands(op, left, right, line);
            return left < right;

          case "<=":
            this.checkNumberOperands(op, left, right, line);
            return left <= right;

          case ">":
            this.checkNumberOperands(op, left, right, line);
            return left > right;

          case ">=":
            this.checkNumberOperands(op, left, right, line);
            return left >= right;

          case "==":
            return left === right;

          case "!=":
            return left !== right;
        }
      }
    }

    throw new Error(`Runtime Error at line ${expr.line}: Cannot evaluate expression`);
  }

  private checkNumberOperands(op: string, left: any, right: any, line: number): void {
    if (typeof left === "number" && typeof right === "number") return;
    throw new Error(`Runtime Error at line ${line}: Operands for '${op}' must be numbers`);
  }

  private isTruthy(val: any): boolean {
    if (val === null || val === undefined) return false;
    if (typeof val === "boolean") return val;
    if (typeof val === "number") return val !== 0;
    return Boolean(val);
  }

  private stringify(val: any): string {
    if (val === null || val === undefined) return "null";
    if (typeof val === "boolean") {
      return val ? "sahi" : "galat";
    }
    return String(val);
  }
}
